src/rust/ide/src/transport/web.rs

```diff
@@ -98,60 +103,198 @@ impl State {
+        on_open.clear_callback();
+        on_close_internal.clear_callback()
+    }
+
+    /// Establish a new WS connection, using the same URL as the previous one.
+    /// All callbacks will be transferred to the new connection.
+    pub fn reconnect(&mut self) -> Result<(),JsValue> {
+       if !self.reconnect {
+           return Err(js_sys::Error::new("Reconnecting has been disabled").into());
+       }
+
+        let url = self.ws.url();
+        warning!(self.logger, "Reconnecting WS to {url}");
+
+        let new_ws = web_sys::WebSocket::new(&url)?;
+
+        self.on_close.         set_target(&new_ws);
+        self.on_error.         set_target(&new_ws);
+        self.on_message.       set_target(&new_ws);
+        self.on_open.          set_target(&new_ws);
+        self.on_close_internal.set_target(&new_ws);
+        self.ws = new_ws;
+
+        Ok(())
+    }
+}
+
+impl Drop for Model {
+    fn drop(&mut self) {
+        info!(self.logger, "Dropping WS model");
+        if let Err(e) = self.close("Rust Value has been dropped.") {
+            error!(self.logger,"Error when closing socket due to being dropped: {js_to_string(&e)}")
+        }
+    }
+}
+
+
+
+// =================
+// === WebSocket ===
+// =================
+
+/// Wrapper over JS `WebSocket` meant for general use.
+#[derive(Clone,CloneRef,Debug)]
+pub struct WebSocket {
+    #[allow(missing_docs)]
+    pub logger : Logger,
+    model : Rc<RefCell<Model>>,
+}
+
+impl WebSocket {
+    /// Wrap given raw JS WebSocket object.
+    pub fn new(ws:web_sys::WebSocket, parent:impl AnyLogger) -> WebSocket {
+        let logger = Logger::sub(parent,ws.url());
+        let model  = Rc::new(RefCell::new(Model::new(ws,logger.clone())));
+        WebSocket {logger,model}
+    }
+
     /// Establish connection with endpoint defined by the given URL and wrap it.
     /// Asynchronous, because it waits until connection is established.
     pub async fn new_opened
-    (parent:impl AnyLogger, url:impl Str) -> Result<WebSocket,ConnectingError> {
-        let ws = web_sys::WebSocket::new(url.as_ref()).map_err(|e| {
-            ConnectingError::ConstructionError(js_to_string(e))
-        })?;
-        let mut wst = WebSocket::new(ws,&parent,url.into());
+    (parent:impl AnyLogger, url:&str) -> Result<WebSocket,ConnectingError> {
+        let ws      = web_sys::WebSocket::new(url).map_err(ConnectingError::construction_error)?;
+        let mut wst = WebSocket::new(ws,&parent);
         wst.wait_until_open().await?;
         Ok(wst)
     }
 
+    /// Generate a callback to be invoked when socket needs reconnecting.
+    fn reconnect_trigger(&self) -> impl FnMut(web_sys::CloseEvent) {
+        let model  = Rc::downgrade(&self.model);
+        let logger = self.logger.clone();
+        move |_| {
+            if let Some(model) = model.upgrade() {
+                if let Err(e) = model.borrow_mut().reconnect() {
+                    error!(logger,"Failed to reconnect: {js_to_string(&e)}");
```

Comment:
missing dot

Reply:
The dot is expected from the pretty-printed JS error.